src/app/forms/services/dynamic-form.types.ts

```diff
@@ -160,6 +164,44 @@ export class CustomFormControl extends FormControl implements CustomControl {
     super(formState, validatorOrOpts, asyncValidator);
     this.meta = meta;
   }
+
+  addWarningValidators(fn: ValidatorFn) {
+    this.warningValidators = [...this.warningValidators, fn];
+  }
+  addAsyncWarningValidators(fn: AsyncValidatorFn) {
+    this.asyncWarningValidators = [...this.asyncWarningValidators, fn];
+  }
+
+  runWarningValidators(): void {
+    const errors: ValidationErrors[] = [];
+
+    this.warningValidators.forEach(validator => {
+      const error = validator(this);
+      if (error) {
+        errors.push(error);
+      }
+    });
+
+    this.asyncWarningValidators.forEach(async validator => {
+      const promiseOrObservable = validator(this);
+      const error = promiseOrObservable instanceof Promise ? await promiseOrObservable : await firstValueFrom(promiseOrObservable);
+      if (error) {
+        errors.push(error);
+      }
+    });
+
+    if (!errors.length) {
+      this.meta.warning = null;
+    } else {
+      this.meta.warning = errors;
+    }
+  }
+}
```

Comment:
These methods are repeated 3 times, once in the `CustomFormControl`, `CustomFormArray` and `CustomFormGroup`. This is mostly because classes can only extend one class.

There is a way using `bind` to reduce this duplication of code. We have used this successfully for `getCleanValue` function in this file.